### PR TITLE
*: encode the returned data using the Chunk format in mocktikv

### PR DESCRIPTION
**Pull request by wshwsh12**
### What problem does this PR solve? 

Return chunk-encoded data when the requested `EncodeType` is `TypeArrow`.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change

Side effects

 - Increased code complexity

### BenchMark

Now I only test the time cost in TPCH-Q5. (10g data)
The reuslt: (Default concurrency)
```
Default  37.11 sec
Arrow    37.67 sec
```
The reuslt: (`set tidb_index_lookup_join_concurrency=16;`)
```
Default  22.80 sec
Arrow    16.68 sec
```
#### Default 
```
+--------------------------------------------------+-------------+------+-------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------+-----------------------+
| id                                               | count       | task | operator info                                                                             | execution info                                                                                                                              | memory                |
+--------------------------------------------------+-------------+------+-------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------+-----------------------+
| Sort_23                                          | 5.00        | root | Column#51:desc                                                                            | time:22.770147694s, loops:2, rows:5                                                                                                         | 1.91796875 KB         |
| └─Projection_25                                  | 5.00        | root | Column#43, Column#49                                                                      | time:22.770131654s, loops:2, rows:5                                                                                                         | N/A                   |
|   └─HashAgg_28                                   | 5.00        | root | group by:Column#52, funcs:sum(Column#0), firstrow(Column#0)                               | time:22.770128546s, loops:2, rows:5                                                                                                         | N/A                   |
|     └─Projection_80                              | 2236833.33  | root | mul(Column#23, minus(1, Column#24)), Column#43, Column#43                                 | time:22.76982236s, loops:76, rows:72985                                                                                                     | N/A                   |
|       └─HashLeftJoin_36                          | 2236833.33  | root | inner join, inner:TableReader_78, equal:[eq(Column#38, Column#4) eq(Column#10, Column#1)] | time:22.769594411s, loops:76, rows:72985                                                                                                    | 23.561248779296875 MB |
|         ├─IndexMergeJoin_47                      | 2236833.33  | root | inner join, inner:TableReader_45, outer key:Column#18, inner key:Column#9                 | time:22.459100187s, loops:1790, rows:1823514                                                                                                | 2.2937463484704494 GB |
|         │ ├─HashRightJoin_51                     | 12006815.85 | root | inner join, inner:HashRightJoin_53, equal:[eq(Column#35, Column#20)]                      | time:19.661978759s, loops:11745, rows:12022687                                                                                              | 1.5376052856445312 MB |
|         │ │ ├─HashRightJoin_53                   | 20000.00    | root | inner join, inner:HashRightJoin_64, equal:[eq(Column#42, Column#38)]                      | time:142.30407ms, loops:23, rows:20042                                                                                                      | 2.05859375 KB         |
|         │ │ │ ├─HashRightJoin_64                 | 5.00        | root | inner join, inner:TableReader_69, equal:[eq(Column#46, Column#44)]                        | time:106.428µs, loops:2, rows:5                                                                                                             | 20.46484375 KB        |
|         │ │ │ │ ├─TableReader_69                 | 1.00        | root | data:Selection_68                                                                         | time:13.526µs, loops:2, rows:1, rpc time:1.092075ms                                                                                         | 169 Bytes             |
|         │ │ │ │ │ └─Selection_68                 | 1.00        | cop  | eq(Column#47, "MIDDLE EAST")                                                              | time:0s, loops:1, rows:1                                                                                                                    | N/A                   |
|         │ │ │ │ │   └─TableScan_67               | 5.00        | cop  | table:region, range:[-inf,+inf], keep order:false                                         | time:0s, loops:1, rows:5                                                                                                                    | N/A                   |
|         │ │ │ │ └─TableReader_66                 | 25.00       | root | data:TableScan_65                                                                         | time:27.713µs, loops:2, rows:25, rpc time:1.274225ms                                                                                        | 688 Bytes             |
|         │ │ │ │   └─TableScan_65                 | 25.00       | cop  | table:nation, range:[-inf,+inf], keep order:false                                         | time:1ms, loops:1, rows:25                                                                                                                  | N/A                   |
|         │ │ │ └─TableReader_71                   | 100000.00   | root | data:TableScan_70                                                                         | time:142.651099ms, loops:99, rows:100000, rpc time:157.204305ms                                                                             | 1.050323486328125 MB  |
|         │ │ │   └─TableScan_70                   | 100000.00   | cop  | table:supplier, range:[-inf,+inf], keep order:false                                       | time:146ms, loops:102, rows:100000                                                                                                          | N/A                   |
|         │ │ └─TableReader_73                     | 59986052.00 | root | data:TableScan_72                                                                         | time:22.233397247s, loops:58582, rows:59986052, rpc max:1.160607429s, min:79.726388ms, avg:661.569041ms, p80:851.602895ms, p95:1.026335118s | 489.6857976913452 MB  |
|         │ │   └─TableScan_72                     | 59986052.00 | cop  | table:lineitem, range:[-inf,+inf], keep order:false                                       | proc max:1.07s, min:73ms, p80:767ms, p95:949ms, rows:59986052, iters:59231, tasks:153                                                       | N/A                   |
|         │ └─TableReader_45                       | 0.15        | root | data:Selection_44                                                                         | time:2m28.076393464s, loops:1893, rows:1249432, rpc max:646.915632ms, min:1.048567ms, avg:274.856929ms, p80:318.035459ms, p95:424.753119ms  | N/A                   |
|         │   └─Selection_44                       | 0.15        | cop  | ge(Column#13, 1994-01-01 00:00:00.000000), lt(Column#13, 1995-01-01)                      | proc max:595ms, min:1ms, p80:312ms, p95:411ms, rows:1249432, iters:10709, tasks:564                                                         | N/A                   |
|         │     └─TableScan_43                     | 0.19        | cop  | table:orders, range: decided by [Column#18], keep order:true                              | proc max:585ms, min:1ms, p80:308ms, p95:401ms, rows:8236741, iters:10709, tasks:564                                                         | N/A                   |
|         └─TableReader_78                         | 1500000.00  | root | data:TableScan_77                                                                         | time:612.626797ms, loops:1466, rows:1500000, rpc max:640.497105ms, min:404.915325ms, avg:534.187743ms, p80:640.497105ms, p95:640.497105ms   | 12.281054496765137 MB |
|           └─TableScan_77                         | 1500000.00  | cop  | table:customer, range:[-inf,+inf], keep order:false                                       | proc max:598ms, min:381ms, p80:598ms, p95:598ms, rows:1500000, iters:1482, tasks:4                                                          | N/A                   |
+--------------------------------------------------+-------------+------+-------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------+-----------------------+
23 rows in set (22.80 sec)
```
image

#### Arrow
```
+--------------------------------------------------+-------------+------+-------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------+-----------------------+
| id                                               | count       | task | operator info                                                                             | execution info                                                                                                                              | memory                |
+--------------------------------------------------+-------------+------+-------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------+-----------------------+
| Sort_23                                          | 5.00        | root | Column#51:desc                                                                            | time:16.672479712s, loops:2, rows:5                                                                                                         | 1.91796875 KB         |
| └─Projection_25                                  | 5.00        | root | Column#43, Column#49                                                                      | time:16.672467204s, loops:2, rows:5                                                                                                         | N/A                   |
|   └─HashAgg_28                                   | 5.00        | root | group by:Column#52, funcs:sum(Column#0), firstrow(Column#0)                               | time:16.672464934s, loops:2, rows:5                                                                                                         | N/A                   |
|     └─Projection_80                              | 2236833.33  | root | mul(Column#23, minus(1, Column#24)), Column#43, Column#43                                 | time:16.671955743s, loops:75, rows:72985                                                                                                    | N/A                   |
|       └─HashLeftJoin_36                          | 2236833.33  | root | inner join, inner:TableReader_78, equal:[eq(Column#38, Column#4) eq(Column#10, Column#1)] | time:16.671436712s, loops:75, rows:72985                                                                                                    | 8.425058089196682 GB  |
|         ├─IndexMergeJoin_47                      | 2236833.33  | root | inner join, inner:TableReader_45, outer key:Column#18, inner key:Column#9                 | time:16.53063965s, loops:1790, rows:1823514                                                                                                 | 3.796726867556572 GB  |
|         │ ├─HashRightJoin_51                     | 12006815.85 | root | inner join, inner:HashRightJoin_53, equal:[eq(Column#35, Column#20)]                      | time:2.943592967s, loops:11746, rows:12022687                                                                                               | 1.5356521606445312 MB |
|         │ │ ├─HashRightJoin_53                   | 20000.00    | root | inner join, inner:HashRightJoin_64, equal:[eq(Column#42, Column#38)]                      | time:166.297341ms, loops:23, rows:20042                                                                                                     | 2.05859375 KB         |
|         │ │ │ ├─HashRightJoin_64                 | 5.00        | root | inner join, inner:TableReader_69, equal:[eq(Column#46, Column#44)]                        | time:615.791µs, loops:2, rows:5                                                                                                             | 620 Bytes             |
|         │ │ │ │ ├─TableReader_69                 | 1.00        | root | data:Selection_68                                                                         | time:18.693µs, loops:2, rows:1, rpc time:868.274µs                                                                                          | 198 Bytes             |
|         │ │ │ │ │ └─Selection_68                 | 1.00        | cop  | eq(Column#47, "MIDDLE EAST")                                                              | time:0s, loops:1, rows:1                                                                                                                    | N/A                   |
|         │ │ │ │ │   └─TableScan_67               | 5.00        | cop  | table:region, range:[-inf,+inf], keep order:false                                         | time:0s, loops:1, rows:5                                                                                                                    | N/A                   |
|         │ │ │ │ └─TableReader_66                 | 25.00       | root | data:TableScan_65                                                                         | time:579.627µs, loops:2, rows:25, rpc time:1.516286ms                                                                                       | 994 Bytes             |
|         │ │ │ │   └─TableScan_65                 | 25.00       | cop  | table:nation, range:[-inf,+inf], keep order:false                                         | time:1ms, loops:1, rows:25                                                                                                                  | N/A                   |
|         │ │ │ └─TableReader_71                   | 100000.00   | root | data:TableScan_70                                                                         | time:161.002089ms, loops:103, rows:100000, rpc time:161.586458ms                                                                            | 1.52813720703125 MB   |
|         │ │ │   └─TableScan_70                   | 100000.00   | cop  | table:supplier, range:[-inf,+inf], keep order:false                                       | time:149ms, loops:102, rows:100000                                                                                                          | N/A                   |
|         │ │ └─TableReader_73                     | 59986052.00 | root | data:TableScan_72                                                                         | time:1.04352478s, loops:59232, rows:59986052, rpc max:1.656616424s, min:137.145763ms, avg:1.200088653s, p80:1.348882139s, p95:1.487314887s  | 1.3459740281105042 GB |
|         │ │   └─TableScan_72                     | 59986052.00 | cop  | table:lineitem, range:[-inf,+inf], keep order:false                                       | proc max:1.168s, min:96ms, p80:904ms, p95:1.003s, rows:59986052, iters:59231, tasks:153                                                     | N/A                   |
|         │ └─TableReader_45                       | 0.15        | root | data:Selection_44                                                                         | time:3m53.824442702s, loops:11285, rows:1249439, rpc max:887.070028ms, min:1.269268ms, avg:427.577992ms, p80:597.372658ms, p95:709.585346ms | N/A                   |
|         │   └─Selection_44                       | 0.15        | cop  | ge(Column#13, 1994-01-01 00:00:00.000000), lt(Column#13, 1995-01-01)                      | proc max:593ms, min:0s, p80:486ms, p95:561ms, rows:1249439, iters:10818, tasks:591                                                          | N/A                   |
|         │     └─TableScan_43                     | 0.19        | cop  | table:orders, range: decided by [Column#18], keep order:true                              | proc max:585ms, min:0s, p80:479ms, p95:553ms, rows:8236791, iters:10818, tasks:591                                                          | N/A                   |
|         └─TableReader_78                         | 1500000.00  | root | data:TableScan_77                                                                         | time:438.678769ms, loops:1483, rows:1500000, rpc max:607.694099ms, min:409.775717ms, avg:508.329883ms, p80:607.694099ms, p95:607.694099ms   | 11.914450645446777 MB |
|           └─TableScan_77                         | 1500000.00  | cop  | table:customer, range:[-inf,+inf], keep order:false                                       | proc max:535ms, min:370ms, p80:535ms, p95:535ms, rows:1500000, iters:1482, tasks:4                                                          | N/A                   |
+--------------------------------------------------+-------------+------+-------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------+-----------------------+
23 rows in set (16.68 sec)
```

image

The cpu time of `Decode` disappears.

I will add more performance test and count the memory use.

### CODE REVIEW

File: distsql/select_result.go [hunk 1]
```diff
@@ -145,6 +178,27 @@ func (r *selectResult) NextRaw(ctx context.Context) (data []byte, err error) {
 // Next reads data to the chunk.
 func (r *selectResult) Next(ctx context.Context, chk *chunk.Chunk) error {
 	chk.Reset()
+	// Check the returned data is default/arrow format.
+	if r.selectResp == nil || len(r.selectResp.RowBatchData) == 0 {
+		err := r.getSelectResp()
+		if err != nil || r.selectResp == nil {
+			return errors.Trace(err)
```
Comment (zz-jason) [inline distsql/select_result.go]: no need to trace anymore.

File: distsql/select_result.go [hunk 2]
```diff
@@ -145,6 +178,27 @@ func (r *selectResult) NextRaw(ctx context.Context) (data []byte, err error) {
...
+		}
+		if len(r.selectResp.RowBatchData) == 0 {
+			r.decodeType = DecodeTypeDefault
```
Comment (zz-jason) [inline distsql/select_result.go]: should we add a metrics here when the expected decode type is `arrow` but the actual decode type is `default`?

File: distsql/select_result.go [hunk 3]
```diff
@@ -163,6 +217,24 @@ func (r *selectResult) Next(ctx context.Context, chk *chunk.Chunk) error {
...
+		err := r.getSelectResp()
+		if err != nil || r.selectResp == nil {
+			return errors.Trace(err)
+		}
+	}
+	r.readRowBatch(chk)
+	return nil
+}
+
+func (r *selectResult) readRowBatch(chk *chunk.Chunk) {
```
Comment (zz-jason) [inline distsql/select_result.go]: I think we don't need to extract a function for this. It's very simple, and the caller is very simple as well.

File: expression/distsql_builtin.go [hunk 4]
```diff
@@ -28,7 +28,8 @@ import (
 	"github.com/pingcap/tipb/go-tipb"
 )
 
-func pbTypeToFieldType(tp *tipb.FieldType) *types.FieldType {
+// PbTypeToFieldType convert tipb.FieldType to FieldType
```
Comment (zz-jason) [inline expression/distsql_builtin.go]:
```suggestion
// PbTypeToFieldType converts tipb.FieldType to FieldType
```

File: session/session_test.go [hunk 5]
```diff
@@ -2018,7 +2018,7 @@ func (s *testSchemaSuite) TestTableReaderChunk(c *C) {
 	}
 	c.Assert(count, Equals, 100)
 	// FIXME: revert this result to new group value after distsql can handle initChunkSize.
-	c.Assert(numChunks, Equals, 1)
+	c.Assert(numChunks, Equals, 10)
```
Comment (zz-jason) [inline session/session_test.go]: Why it's changed to `10`?

File: util/chunk/codec.go [hunk 6]
```diff
@@ -105,6 +105,7 @@ func (c *Codec) DecodeToChunk(buffer []byte, chk *Chunk) (remained []byte) {
 	return buffer
 }
 
+// Todo: Optimize all data is null.
```
Comment (zz-jason) [inline util/chunk/codec.go]: Comments should be started with the function name in golang.

File: session/session_test.go (hunk 5 above)
Comment (wshwsh12) [reply session/session_test.go]: In the test case, it `SplitTable` to 10 regions and will get 10 chunks in response.
In default format, it will append all result to one chunk.
In arrow format, it will return the chunk directly. So `numChunks` is changed to 10.

File: expression/distsql_builtin.go (hunk 4 above)
Comment (wshwsh12) [reply expression/distsql_builtin.go]: Update.

File: distsql/select_result.go (hunk 3 above)
Comment (wshwsh12) [reply distsql/select_result.go]: Update.

File: distsql/select_result.go (hunk 1 above)
Comment (wshwsh12) [reply distsql/select_result.go]: Update.

File: distsql/select_result.go [hunk 7]
```diff
@@ -38,10 +38,41 @@ var (
 	_ SelectResult = (*streamResult)(nil)
 )
 
+// DecodeType indicates the encode type.
```
[the remark this answers is not in the document]
Comment (wshwsh12) [reply distsql/select_result.go]: OK. I remove the field `DecodeType` and use `tipb.EncodeType`.

File: distsql/distsql.go [hunk 8]
```diff
@@ -141,3 +144,70 @@ func Checksum(ctx context.Context, client kv.Client, kvReq *kv.Request, vars *kv
 	}
 	return result, nil
 }
+
+// SetEncodeType set a DAGRequest's EncodeType
```
[the remark this answers is not in the document]
Comment (wshwsh12) [reply distsql/distsql.go]: Update.

File: distsql/distsql.go [hunk 9]
```diff
@@ -141,3 +144,70 @@ func Checksum(ctx context.Context, client kv.Client, kvReq *kv.Request, vars *kv
...
+						return false
+					}
+				}
+			}
+		}
+	}
+	return true
+}
+
+func typeSupported(tpInput int32) bool {
```
[the remark this answers is not in the document]
Comment (wshwsh12) [reply distsql/distsql.go]: Update.

File: distsql/select_result.go (hunk 2 above)
Comment (wshwsh12) [reply distsql/select_result.go]: Ok.

File: util/chunk/codec.go (hunk 6 above)
Comment (wshwsh12) [reply util/chunk/codec.go]: Update.

**Comment by wshwsh12** [comment]
Update. PTAL again. @zz-jason

File: distsql/select_result.go [hunk 10]
```diff
@@ -145,6 +147,28 @@ func (r *selectResult) NextRaw(ctx context.Context) (data []byte, err error) {
...
+		}
+	}
+
+	switch r.encodeType {
+	case tipb.EncodeType_TypeDefault:
+		return r.readFromDefault(ctx, chk)
+	case tipb.EncodeType_TypeArrow:
+		return r.readFromArrow(ctx, chk)
+	}
+	panic("unsupported decode type")
```
Comment (zz-jason) [inline distsql/select_result.go]:
```suggestion
	return errors.Errorf("unsupported encode type:%v", r.encodeType)
```

File: distsql/select_result.go [hunk 11]
```diff
@@ -145,6 +147,28 @@ func (r *selectResult) NextRaw(ctx context.Context) (data []byte, err error) {
...
+}
+
+func (r *selectResult) readFromDefault(ctx context.Context, chk *chunk.Chunk) error {
 	for !chk.IsFull() {
 		if r.selectResp == nil || r.respChkIdx == len(r.selectResp.Chunks) {
```
Comment (zz-jason) [inline distsql/select_result.go]: no need to do this check again.

File: distsql/select_result.go [hunk 12]
```diff
@@ -163,6 +187,20 @@ func (r *selectResult) Next(ctx context.Context, chk *chunk.Chunk) error {
 	return nil
 }
 
+func (r *selectResult) readFromArrow(ctx context.Context, chk *chunk.Chunk) error {
+	if r.selectResp == nil || len(r.selectResp.RowBatchData) == 0 {
```
Comment (zz-jason) [inline distsql/select_result.go]: ditto

File: distsql/select_result.go [hunk 13]
```diff
@@ -163,6 +187,20 @@ func (r *selectResult) Next(ctx context.Context, chk *chunk.Chunk) error {
...
+		err := r.getSelectResp()
+		if err != nil || r.selectResp == nil {
+			return errors.Trace(err)
+		}
+	}
+	rowBatchData := r.selectResp.RowBatchData
+	codec := chunk.NewCodec(r.fieldTypes)
+	remained := codec.DecodeToChunk(rowBatchData, chk)
```
Comment (zz-jason) [inline distsql/select_result.go]: we should put multiple `rowBatchData` into one `chk` in order to make it full.

File: executor/analyze.go [hunk 14]
```diff
@@ -270,7 +270,7 @@ func (e *AnalyzeIndexExec) fetchAnalyzeResult(ranges []*ranger.Range, isNullRang
 	if err != nil {
 		return err
 	}
-	result.Fetch(ctx)
+	result.Fetch(ctx, tipb.EncodeType_TypeDefault)
```
Comment (zz-jason) [inline executor/analyze.go]: Would it be faster if we support chunk encoded data for the `Analyze` API?

File: executor/checksum.go [hunk 15]
```diff
@@ -132,7 +132,7 @@ func (e *ChecksumTableExec) handleChecksumRequest(req *kv.Request) (resp *tipb.C
 	if err != nil {
 		return nil, err
 	}
-	res.Fetch(ctx)
+	res.Fetch(ctx, tipb.EncodeType_TypeDefault)
```
Comment (zz-jason) [inline executor/checksum.go]: ditto

File: executor/table_readers_required_rows_test.go [hunk 16]
```diff
@@ -16,6 +16,7 @@ package executor
 import (
 	"context"
 	"fmt"
+	"github.com/pingcap/tidb/planner/core"
```
Comment (zz-jason) [inline executor/table_readers_required_rows_test.go]: please reorg the imported libraries.

File: executor/analyze.go (hunk 14 above)
Comment (zz-jason) [reply executor/analyze.go]: We can open an issue for this, leave it to the future optimizations.

File: distsql/select_result.go (hunk 10 above)
Comment (wshwsh12) [reply distsql/select_result.go]: Update.

File: distsql/select_result.go (hunk 13 above)
Comment (wshwsh12) [reply distsql/select_result.go]: There are many details to judge `!chk.IsFull()` in arrow format.  Need a struct to maintain the `remained` . 
 I will do this in the next pr.

File: executor/table_readers_required_rows_test.go (hunk 16 above)
Comment (wshwsh12) [reply executor/table_readers_required_rows_test.go]: Update.

File: distsql/select_result.go (hunk 11 above)
Comment (wshwsh12) [reply distsql/select_result.go]: Update.  `r.selectResp == nil` is unnecessary, but `r.respChkIdx == len(r.selectResp.Chunks)` is need.

File: distsql/select_result.go (hunk 12 above)
Comment (wshwsh12) [reply distsql/select_result.go]: OK. I remove the check.

File: executor/analyze.go (hunk 14 above)
Comment (wshwsh12) [reply executor/analyze.go]: OK.

File: distsql/select_result.go (hunk 13 above)
Comment (zz-jason) [reply distsql/select_result.go]: okay

File: distsql/select_result.go [hunk 17]
```diff
@@ -145,12 +147,35 @@ func (r *selectResult) NextRaw(ctx context.Context) (data []byte, err error) {
...
 func (r *selectResult) Next(ctx context.Context, chk *chunk.Chunk) error {
 	chk.Reset()
+	// Check the returned data is default/arrow format.
+	if r.selectResp == nil || (len(r.selectResp.RowBatchData) == 0 && r.respChkIdx == len(r.selectResp.Chunks)) {
+		err := r.getSelectResp()
+		if err != nil || r.selectResp == nil {
+			return err
+		}
+		if len(r.selectResp.RowBatchData) == 0 {
+			r.encodeType = tipb.EncodeType_TypeDefault
```
Comment (zz-jason) [inline distsql/select_result.go]: IMHO, we only need to do this check when the `Next()` function is called the first time. All the cop responses should have the same decode method. Correct me if I'm wrong.

Comment (wshwsh12) [reply distsql/select_result.go]: You are right.

Comment (wshwsh12) [reply distsql/select_result.go]: Now the check is temporary ad-hoc. The next pr will add a field `response_format` in response, and use the field to decide the decode format. Then i will remove the check in next pr. It is OK?

File: distsql/distsql.go [hunk 18]
```diff
@@ -141,3 +144,74 @@ func Checksum(ctx context.Context, client kv.Client, kvReq *kv.Request, vars *kv
...
+}
+
+func useChunkIPC(dagReq *tipb.DAGRequest) bool {
+	if config.GetGlobalConfig().TiKVClient.EnableArrow == false {
+		return false
+	}
+	if config.GetGlobalConfig().EnableStreaming == true {
+		return false
+	}
+	root := dagReq.Executors[len(dagReq.Executors)-1]
```
Comment (XuHuaiyu) [inline distsql/distsql.go]: The checks for root is not necessary, cause it has been checked before SetEncodeType

File: distsql/distsql.go [hunk 19]
```diff
@@ -141,3 +144,74 @@ func Checksum(ctx context.Context, client kv.Client, kvReq *kv.Request, vars *kv
...
+					}
+				}
+			}
+		}
+	}
+	return true
+}
+
+// typeSupported indicate the type support to encode in arrow format.
+func typeSupported(tpInput int32) bool {
```
Comment (XuHuaiyu) [inline distsql/distsql.go]: Useless function.

File: distsql/select_result.go [hunk 20]
```diff
@@ -41,7 +41,7 @@ var (
 // SelectResult is an iterator of coprocessor partial results.
 type SelectResult interface {
 	// Fetch fetches partial results from client.
-	Fetch(context.Context)
+	Fetch(context.Context, tipb.EncodeType)
```
Comment (XuHuaiyu) [inline distsql/select_result.go]: We can put EncodeType in SessionVars and fetch it when building the selectResult.

File: store/mockstore/mocktikv/cop_handler_dag.go [hunk 21]
```diff
@@ -550,13 +551,111 @@ func (mock *mockCopStreamClient) readBlockFromExecutor() (tipb.Chunk, bool, *cop
...
+	case tipb.EncodeType_TypeArrow:
+		err := h.encodeArrow(selResp, rows, colTypes, dagReq.OutputOffsets, loc)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func (h *rpcHandler) constructRespSchema(dagCtx *dagContext) []*types.FieldType {
```
Comment (XuHuaiyu) [inline store/mockstore/mocktikv/cop_handler_dag.go]: Is this function also useless, dagCtx.evalCtx.fieldTps is enough?

File: store/mockstore/mocktikv/cop_handler_dag.go [hunk 22]
```diff
@@ -550,13 +551,111 @@ func (mock *mockCopStreamClient) readBlockFromExecutor() (tipb.Chunk, bool, *cop
...
 		OutputCounts: counts,
 	}
+	for i := range warnings {
+		selResp.Warnings = append(selResp.Warnings, toPBError(warnings[i].Err))
+	}
+
+	return selResp
+}
+
+func (h *rpcHandler) supplementSelectResponse(selResp *tipb.SelectResponse, dagReq *tipb.DAGRequest, dagCtx *dagContext, rows [][][]byte) error {
```
Comment (XuHuaiyu) [inline store/mockstore/mocktikv/cop_handler_dag.go]: s/ supplementSelectResponse/ fillUpData4SelectResponse

File: store/mockstore/mocktikv/cop_handler_dag.go [hunk 23]
```diff
@@ -76,20 +74,23 @@ func (h *rpcHandler) handleCopDAGRequest(req *coprocessor.Request) *coprocessor.
...
 	var execDetails []*execDetail
 	if dagReq.CollectExecutionSummaries != nil && *dagReq.CollectExecutionSummaries {
 		execDetails = e.ExecDetails()
 	}
-	return buildResp(chunks, e.Counts(), execDetails, err, warnings)
+
+	selResp := h.initSelectResponse(err, dagCtx.evalCtx.sc.GetWarnings(), e.Counts())
+	if err == nil {
+		err = h.supplementSelectResponse(selResp, dagReq, dagCtx, rows)
+		if err != nil {
```
Comment (XuHuaiyu) [inline store/mockstore/mocktikv/cop_handler_dag.go]: Do we need to log the error here? It'll be returned by the response.

File: store/mockstore/mocktikv/cop_handler_dag.go [hunk 24]
```diff
@@ -550,13 +551,111 @@ func (mock *mockCopStreamClient) readBlockFromExecutor() (tipb.Chunk, bool, *cop
...
+	respColTypes := make([]*types.FieldType, 0, len(colOrdinal))
+	for _, ordinal := range colOrdinal {
+		respColTypes = append(respColTypes, colTypes[ordinal])
+	}
+
+	chk := chunk.NewChunkWithCapacity(respColTypes, rowsPerChunk)
+	encoder := chunk.NewCodec(respColTypes)
+	decoder := codec.NewDecoder(chk, loc)
+	for i := range rows {
+		for j, ordinal := range colOrdinal {
```
Comment (XuHuaiyu) [inline store/mockstore/mocktikv/cop_handler_dag.go]: for j := range respColTypes {
    s/ colTypes[ordinal]/ respColTypes[j]
}

File: store/mockstore/mocktikv/cop_handler_dag.go [hunk 25]
```diff
@@ -550,13 +548,113 @@ func (mock *mockCopStreamClient) readBlockFromExecutor() (tipb.Chunk, bool, *cop
...
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func (h *rpcHandler) constructRespSchema(dagCtx *dagContext) []*types.FieldType {
+	root := dagCtx.dagReq.Executors[len(dagCtx.dagReq.Executors)-1]
+	if root.Aggregation != nil {
```
Comment (XuHuaiyu) [inline store/mockstore/mocktikv/cop_handler_dag.go]:
``` go
agg := root.Aggregation
if root.StreamAgg != nil {
   agg = root.StreamAgg
}
if agg == nil {
    return dagCtx.evalCtx.fieldTps
}
...
```

File: store/mockstore/mocktikv/cop_handler_dag.go [hunk 26]
```diff
@@ -550,13 +548,113 @@ func (mock *mockCopStreamClient) readBlockFromExecutor() (tipb.Chunk, bool, *cop
...
 	}
+	for i := range warnings {
+		selResp.Warnings = append(selResp.Warnings, toPBError(warnings[i].Err))
+	}
+
+	return selResp
+}
+
+func (h *rpcHandler) fillUpData4SelectResponse(selResp *tipb.SelectResponse, dagReq *tipb.DAGRequest, dagCtx *dagContext, rows [][][]byte) error {
+	colTypes := h.constructRespSchema(dagCtx)
```
Comment (XuHuaiyu) [inline store/mockstore/mocktikv/cop_handler_dag.go]: line 564 and line 565 can be moved after line 570

File: store/mockstore/mocktikv/cop_handler_dag.go [hunk 27]
```diff
@@ -550,13 +548,113 @@ func (mock *mockCopStreamClient) readBlockFromExecutor() (tipb.Chunk, bool, *cop
...
+	loc := dagCtx.evalCtx.sc.TimeZone
+
+	switch dagReq.EncodeType {
+	case tipb.EncodeType_TypeDefault:
+		h.encodeDefault(selResp, rows, dagReq.OutputOffsets)
+	case tipb.EncodeType_TypeArrow:
+		err := h.encodeArrow(selResp, rows, colTypes, dagReq.OutputOffsets, loc)
+		if err != nil {
+			return errors.Trace(err)
```
Comment (XuHuaiyu) [inline store/mockstore/mocktikv/cop_handler_dag.go]: `return err` is enougn

File: store/mockstore/mocktikv/cop_handler_dag.go [hunk 28]
```diff
@@ -550,13 +548,113 @@ func (mock *mockCopStreamClient) readBlockFromExecutor() (tipb.Chunk, bool, *cop
...
+	selResp.Chunks = chunks
+}
+
+func (h *rpcHandler) encodeArrow(selResp *tipb.SelectResponse, rows [][][]byte, colTypes []*types.FieldType, colOrdinal []uint32, loc *time.Location) error {
+	rowBatchData := make([]byte, 0, 1024)
+	respColTypes := make([]*types.FieldType, 0, len(colOrdinal))
+	for _, ordinal := range colOrdinal {
+		respColTypes = append(respColTypes, colTypes[ordinal])
+	}
+
```
Comment (XuHuaiyu) [inline store/mockstore/mocktikv/cop_handler_dag.go]: useless empty line

File: sessionctx/variable/sysvar.go [hunk 29]
```diff
@@ -675,6 +675,7 @@ var defaultSysVars = []*SysVar{
 	{ScopeSession, TIDBMemQuotaIndexLookupJoin, strconv.FormatInt(DefTiDBMemQuotaIndexLookupJoin, 10)},
 	{ScopeSession, TIDBMemQuotaNestedLoopApply, strconv.FormatInt(DefTiDBMemQuotaNestedLoopApply, 10)},
 	{ScopeSession, TiDBEnableStreaming, "0"},
+	{ScopeSession, TiDBEnableArrow, "0"},
```
Comment (winoros) [inline sessionctx/variable/sysvar.go]: We'll set its default value as 0 when this pr is merged?

Comment (wshwsh12) [reply sessionctx/variable/sysvar.go]: Yes.

File: sessionctx/variable/session.go [hunk 30]
```diff
@@ -352,6 +352,9 @@ type SessionVars struct {
 	// TODO: remove this after tidb-server configuration "enable-streaming' removed.
 	EnableStreaming bool
 
+	// EnableArrow indicates whether the coprocessor request can use arrow API.
+	EnableArrow bool
```
Comment (zz-jason) [inline sessionctx/variable/session.go]: We can make this config item be able to be reloaded without restarting the tidb-server like other hot-reloadable config items. It's not urgent, we can do this in another PR.
